### PR TITLE
Update date & time function

### LINKED ISSUE #903: Introduce Improve DateTime Handling from Upstream
## Expected behavior

https://make.wordpress.org/core/2019/09/23/date-time-improvements-wp-5-3/

The behavior is described in this blog post from 2019 and the related fixes.

## Current behavior

Date and Time handling in the existing code is inconsistent at best. The upstream proposal from WordPress would not solve all the problems with it, but it would clean them all up.

## Possible solution

I see no reason why the functions and unit tests could not be incorporated as is from the commits, as they would not break any sort of compatibility, as older functions became wrappers around the newer enhancements.

## Context

I deal a lot in my code with date handling and related timezone issues. This simplifies the consistency of time display immensely.

### PR DESCRIPTION
**Pull request by mattyrob**
## Description
This PR is a backport of a major upgrade to the core date and time functions. Fixes #903

## Motivation and context
The original date and time functions originate back in the days of PHP4. Upstream in WordPress there has been a major piece of work updating these functions to make them more reliable.

## How has this been tested?
This PR include around 60 changeset backports, with each local testing has been done to ensure unit tests are still passing. This enabled several other changes sets to be identified and back ported as they were needed for particular changeset as part of this work. More details of the changes sets reviewed and back ported is given below.

```
43380 - conflict
43387 - clean
43434 - conflict
43562 - clean
43591 - clean
43594 - clean
43599 - null
43610 - null
43703 - null
44592 - clean
44710 - clean
44809 - conflict
45375 - conflict
45377 - clean
45379 - conflict
45424 - conflict
45556 - conflict
45850 - conflict
45853 - conflict
45855 - clean
45856 - clean
45876 - conflict
45882 - conflict, better with 43380
45883 - clean
45887 - conflict
45900 - clean
45902 - clean
46083 - conflict
46569 - clean, better with 45901 and 45914
46578 - conflict, better with 46154
46579 - conflict
46580 - clean
46583 - clean
46584 - clean
46756 - not needed 5.2.0 code
46757 - not needed 5.2.0 code
46774 - not needed 5.2.0 code
46785 - not needed - test file not in CP
46795 - conflict
46796 - clean
46862 - conflict
46863 - not needed, branch
46958 - clean
46973 - conflict
46974 - not needed, patches function introduced in 5.2.0
46977 - not needed, patches function in branch introduced in 5.2.0
46981 - not needed, patches testing of function introduced in 5.2.0
46982 - not needed, patches testing in branch of function introduced in 5.2.0
47073 - conflict
47076 - clean
47078 - conflict
47079 - conflict
47080 - conflict
47098 - conflict
47127 - not needed, backport to branch
48911 - conflict, needs 45908, 45911 and 47118
48912 - conflict, needs 47287
48918 - clean
48920 - not needed, for branch
48921 - not needed, for branch
48922 - not needed, for branch
48924 - conflict
48926 - not needed, for branch
51905 - conflict
51906 - conflict
51950 - conflict
```

## Screenshots
N/A

## Types of changes
- New feature / enhancement

### CODE REVIEW

**Comment by joyously** [comment]
I think this is related (was done 3 days ago):
https://core.trac.wordpress.org/changeset/52586
> WordPress commit r52586
> Date/Time: Fix race conditions in `current_time()` tests.

**Comment by mattyrob** [comment]
52586 is a branch backport of core changeset:
https://core.trac.wordpress.org/changeset/45857

It is still missing from the long list I'd found above so updated PR is to follow, race conditions are always worth fixing.

File: src/wp-admin/includes/meta-boxes.php [hunk 1]
```diff
@@ -193,56 +193,74 @@ function post_submit_meta_box( $post, $args = array() ) {
...
+		<a href="#edit_timestamp" class="edit-timestamp hide-if-no-js" role="button">
+			<span aria-hidden="true"><?php _e( 'Edit' ); ?></span>
+			<span class="screen-reader-text"><?php _e( 'Edit date and time' ); ?></span>
+		</a>
+		<fieldset id="timestampdiv" class="hide-if-js">
+			<legend class="screen-reader-text"><?php _e( 'Date and time' ); ?></legend>
+			<?php touch_time( ( $action === 'edit' ), 1 ); ?>
+		</fieldset>
+		</div>
+		<?php // /misc-pub-section ?>
```
Comment (xxsimoxx) [inline src/wp-admin/includes/meta-boxes.php:260]: The comment and the closing php tag seems to lead to this:
image

Comment (xxsimoxx) [inline src/wp-admin/includes/meta-boxes.php:260]: Looks like the closing PHP tag is wrong.
image

**Review by xxsimoxx — CHANGES_REQUESTED** [review]

File: src/wp-admin/includes/meta-boxes.php (hunk 1 above)
Comment (mattyrob) [reply src/wp-admin/includes/meta-boxes.php]: Good spot, thanks. Fix pushed for this.

Comment (xxsimoxx) [reply src/wp-admin/includes/meta-boxes.php]: Cherry picked. Now it's ok!

**Comment by mattyrob** [comment]
As pre review comment on Slack - thanks @xxsimoxx

**Pull request merged**